reconcilers/flow.go

```diff
@@ -0,0 +1,315 @@
+	Condition func(ctx context.Context, resource Type) bool
+
+	// Reconciler is called so long as Condition() returns true. Typically,
+	// Reconciler is a Sequence of multiple SubReconcilers.
+	Reconciler SubReconciler[Type]
+
+	// MaxIterations guards against infinite loops by limiting the number of
+	// allowed iterations before returning an error. Defaults to 100, set to 0
+	// to disable.
+	MaxIterations *int
+
+	lazyInit sync.Once
+}
+
+func (r *While[T]) init() {
+	r.lazyInit.Do(func() {
+		if r.Name == "" {
+			r.Name = "While"
+		}
+		if r.MaxIterations == nil {
+			r.MaxIterations = pointer.Int(100)
+		}
+	})
+}
+
+func (r *While[T]) validate(ctx context.Context) error {
+	// require If
+	if r.Condition == nil {
+		return fmt.Errorf("While %q must implement Condition", r.Name)
+	}
+
+	// require Reconciler
+	if r.Reconciler == nil {
+		return fmt.Errorf("While %q must implement Reconciler", r.Name)
+	}
+
+	return nil
+}
+
+func (r *While[T]) SetupWithManager(ctx context.Context, mgr ctrl.Manager, bldr *builder.Builder) error {
+	r.init()
+
+	log := logr.FromContextOrDiscard(ctx).
+		WithName(r.Name)
+	ctx = logr.NewContext(ctx, log)
+
+	if err := r.validate(ctx); err != nil {
+		return err
+	}
+
+	if err := r.Reconciler.SetupWithManager(ctx, mgr, bldr); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *While[T]) Reconcile(ctx context.Context, resource T) (Result, error) {
+	r.init()
+
+	log := logr.FromContextOrDiscard(ctx).
+		WithName(r.Name)
+	ctx = logr.NewContext(ctx, log)
+
+	aggregateResult := Result{}
+	for i := 0; r.Condition(ctx, resource); i++ {
+		if *r.MaxIterations != 0 && i >= *r.MaxIterations {
+			// TODO make this a typed error
+			return aggregateResult, fmt.Errorf("exceeded max iterations")
+		}
+
+		log := logr.FromContextOrDiscard(ctx).
+			WithName(fmt.Sprintf("%d", i))
+		ctx := logr.NewContext(ctx, log)
+
+		result, err := r.Reconciler.Reconcile(ctx, resource)
+		aggregateResult = AggregateResults(aggregateResult, result)
+		if err != nil {
+			return aggregateResult, err
+		}
+	}
+
+	return aggregateResult, nil
+}
+
+type TryCatch[Type client.Object] struct {
+	// Name used to identify this reconciler.  Defaults to `TryCatch`.  Ideally
+	// unique, but not required to be so.
+	//
+	// +optional
+	Name string
+
+	// Setup performs initialization on the manager and builder this reconciler
+	// will run with. It's common to setup field indexes and watch resources.
+	//
+	// +optional
+	Setup func(ctx context.Context, mgr Manager, bldr *Builder) error
+
+	// If controls the flow of execution calling the Then subreconciler when
+	// true, and the Else subreconciler when false. Typically, Try is a
+	// Sequence of multiple SubReconcilers.
+	Try SubReconciler[Type]
+
+	// Catch is called with the results from Try(). New results can be returned
+	// suppressing the original results.
+	//
+	// +optional
+	Catch func(ctx context.Context, resource Type, result Result, err error) (Result, error)
```

Comment:
After trying `TryCatch` on for size mentally, I am wondering if an advice-style reconciler could be implemented with it. Say, updated a single condition on a resource depending on whether a collection of steps was successful or not like so:

```go
func ApplyResources(c reconcilers.Config) *reconcilers.SubReconciler {
	return &TryCatch{
		Try: reconcilers.Sequence{
			ApplyRegistryCredentialsPlaceholder(c),
			ApplyCerts(c),
			ApplyRedis(c, images.Redis),
			ApplyAuthServer(c, images.AuthServer),
			ApplyTokenSignatureKeys(c),
		},
		Catch: func(ctx, parent *v1alpha1.AuthServer, result Result, err error) {
			if err != nil {
				parent.Status.MarkResourcesError(ctx, err.Error())
				return result, nil
			} else {
				parent.Status.MarkResourcesApplied(ctx)
				return result, nil
			}
		},
		Config: c,
	}
}
```

Reply:
Exactly. You'd probably want to do a little bit of detection on the err to figure out if it is something that should be suppressed or continued.